daemon/api_download.go

```diff
@@ -68,38 +108,154 @@ func postSnapDownload(c *Command, r *http.Request, user *auth.UserState) Respons
+			return InternalError(err.Error())
+		}
+		ss.stream = r
+		if s != 206 {
+			// store/cdn has no partial content (valid
+			// reply per RFC)
+			logger.Debugf("store refused our range request")
+			ss.resume = 0
+		}
 	}
-	info := sars[0].Info
 
-	downloadInfo := info.DownloadInfo
-	resume := int64(0)
-	r, status, err := getStore(c).DownloadStream(context.TODO(), action.SnapName, &downloadInfo, resume, user)
+	return ss
+}
+
+func newSnapStream(snapName string, info *snap.Info, secret []byte) (*snapStream, error) {
+	dlInfo := &info.DownloadInfo
+	fname := filepath.Base(info.MountFile())
+	tokenJSON := downloadTokenJSON{
+		SnapName: snapName,
+		Filename: fname,
+		Info:     dlInfo,
+	}
+	tokStr, err := sealDownloadToken(&tokenJSON, secret)
 	if err != nil {
-		return InternalError(err.Error())
+		return nil, err
 	}
-	// XXX: check for 206 as well here once we set resume to something other
-	// than "0"
-	if status != 200 {
-		return InternalError("internal error: unexpected status code from DownloadStream: %v", status)
+	return &snapStream{
+		SnapName: snapName,
+		Filename: fname,
+		Info:     dlInfo,
+		Token:    tokStr,
+	}, nil
+}
+
+func newResumingSnapStream(snapName string, tokStr string, secret []byte) (*snapStream, error) {
+	d, err := unsealDownloadToken(tokStr, secret)
+	if err != nil {
+		return nil, err
 	}
+	if d.SnapName != snapName {
+		return nil, fmt.Errorf("resume snap name does not match original snap name")
+	}
+	return &snapStream{
+		SnapName: snapName,
+		Filename: d.Filename,
+		Info:     d.Info,
+	}, nil
+}
+
+type downloadTokenJSON struct {
+	SnapName string             `json:"snap-name"`
+	Filename string             `json:"filename"`
+	Info     *snap.DownloadInfo `json:"dl-info"`
+}
+
+func sealDownloadToken(d *downloadTokenJSON, secret []byte) (string, error) {
+	b, err := json.Marshal(d)
+	if err != nil {
+		return "", err
+	}
+	mac := hmac.New(sha256.New, secret)
+	mac.Write(b)
+	tok := mac.Sum(b)
+	return base64.RawURLEncoding.EncodeToString(tok), nil
+}
 
-	return fileStream{
-		SnapName: action.SnapName,
-		Filename: info.Filename(),
-		Info:     downloadInfo,
-		stream:   r,
+var errInvalidDownloadToken = errors.New("download token is invalid")
+
+func unsealDownloadToken(tokStr string, secret []byte) (*downloadTokenJSON, error) {
+	tok, err := base64.RawURLEncoding.DecodeString(tokStr)
+	if err != nil {
+		return nil, errInvalidDownloadToken
+	}
+	sz := len(tok)
+	if sz < sha256.Size {
+		return nil, errInvalidDownloadToken
+	}
+	h := tok[sz-sha256.Size:]
+	b := tok[:sz-sha256.Size]
+	mac := hmac.New(sha256.New, secret)
+	mac.Write(b)
+	if !hmac.Equal(h, mac.Sum(nil)) {
+		return nil, errInvalidDownloadToken
+	}
+	var d downloadTokenJSON
+	if err := json.Unmarshal(b, &d); err != nil {
+		return nil, err
+	}
+	return &d, nil
+}
+
+func downloadTokensSecret(c *Command) (secret []byte, err error) {
+	st := c.d.overlord.State()
+	st.Lock()
+	defer st.Unlock()
+	const k = "api-download-tokens-secret"
```

Comment:
What is the consequence of using a fixed secret for all downloads? I think this needs to be documented. Out of the whole design this is the most uneasy part I read so far. I would have expected this to be some kind of temporary secret that exists for the purpose of a specific API request chain.

Reply:
Making it different between requests is not very interesting, by design it meant to be hard to guess if not accessible. We could use a new one each time snapd restarts, it means snap download between restarts would get errors, we could rotate each x months, it adds some complexity. We can definitely do it in a follow up though. The main goal here is avoiding casual tampering and keeping integrity, using a HMAC is just as easy/easier than inventing an ad hoc approach, there is no deep security implications because we don't avoid the usual API auth checks even when the token is given from one hand, from another anybody can try to access the store bypassing snapd.